File: frontend/src/types/job.ts
import type { GenerationOptions } from './generationOptions'

export type JobStatus = 'pending' | 'processing' | 'ready' | 'failed'

/** Pipeline stage currently running; `null` while the job is `pending`. */
export type JobStage =
  | 'rendering'
  | 'extracting'
  | 'generating_cards'
  | 'detecting_masks'
  | 'composing'
  | 'finalizing'

export interface Job {
  id: number
  deck_name: string
  status: JobStatus
  error_message: string | null
  page_count: number | null
  card_count: number
  stage: JobStage | null
  /** Human-readable label for `stage`; `null` iff `stage` is `null`. */
  stage_label: string | null
  /**
   * Float in `[0, 100]`, or `null` when the current stage's denominator is
   * unknown. Server-computed — never re-derived or defaulted to `0` here.
   */
  progress_percent: number | null
  /** Integer seconds remaining, or `null` when no honest ETA exists. */
  eta_seconds: number | null
  /**
   * Generation options that produced this job. Always present and fully
   * populated — the server resolves every key at creation time.
   */
  options: GenerationOptions
  created_at: string
  updated_at: string
}

File: backend/app/schemas/job.py
from datetime import datetime

from pydantic import BaseModel

from app.models.enums import JobStage, JobStatus
from app.schemas.generation_options import GenerationOptions


class JobRead(BaseModel):
    id: int
    deck_name: str
    status: JobStatus
    error_message: str | None
    page_count: int | None
    card_count: int
    stage: JobStage | None
    stage_label: str | None
    progress_percent: float | None
    eta_seconds: int | None
    #: The resolved generation options that produced this job; always populated.
    options: GenerationOptions
    created_at: datetime
    updated_at: datetime
